### DISCUSSION TITLE
missing Transport CityObject

### DISCUSSION THREAD

**Discussion by AdrianKriger** [cityjson/cjio#141] [Q&A] [answered]
Hi.

I might not be asking this question in the right place. *---and its possibly an issue with my clunky code.*

Some CityObjects are not being created.

![image](https://user-images.githubusercontent.com/59996720/174576772-04da802b-c95e-4630-ad43-3d5612ddc9d0.png)

Notice the gaps in the terrain where the Transport CityObject (black road) should be.

The process is -> after some basic [vector manipulation](https://github.com/AdrianKriger/osm_LoD1_3DCityModel/blob/main/village_campus/mwe/osm3DMain2.py#L69) -> triangulate each  [section of road](https://github.com/AdrianKriger/osm_LoD1_3DCityModel/blob/main/village_campus/mwe/osm3DMain2.py#L99) -> save these and [pop them in as CityJSON Objects](https://github.com/AdrianKriger/osm_LoD1_3DCityModel/blob/main/village_campus/mwe/osm3DMain2.py#L121). 

The key variables here are [idxALL](https://github.com/AdrianKriger/osm_LoD1_3DCityModel/blob/main/village_campus/mwe/osm3DCode2.py#L847-L853) (the road segments), [t_list](https://github.com/AdrianKriger/osm_LoD1_3DCityModel/blob/main/village_campus/mwe/osm3DCode2.py#L855-L861) (the triangulation) and [rd_pts](https://github.com/AdrianKriger/osm_LoD1_3DCityModel/blob/main/village_campus/mwe/osm3DCode2.py#L837-L845) (the vertices). Which are added as such -> [here](https://github.com/AdrianKriger/osm_LoD1_3DCityModel/blob/main/village_campus/mwe/osm3DCode2.py#L1747-L1758):
```python
def add_rd_v(pts, cm):
    #cm['vertices'] = pts
    for p in pts:
        cm['vertices'].append([p[0], p[1], p[2]])
        #print(cm['vertices'])

def add_rd_b(T, r, acoi, allsurfaces, cm):
    for i in T:
        #allsurfaces.append([[i[0], i[1], i[2]]]) 
        allsurfaces.append([[i[0]+len(cm['vertices'])-len(r)-len(acoi), 
                             i[1]+len(cm['vertices'])-len(r)-len(acoi), 
                             i[2]+len(cm['vertices'])-len(r)-len(acoi)]])
```
I know the triangulation is there. If I execute 
```python
import pyvista as pv
number = 6

l2 = np.vstack(idxAll[number])
l2 = l2.reshape([-1, 2])
twos2 = np.array([[2]] * len(idxAll[number]))
linesR = np.append(twos2, l2, axis=1)
    
trinR = pv.PolyData(rd_pts[number])
polygonR = pv.PolyData(rd_pts[number])
holes = pv.PolyData()
# Make sure it has the same points as the mesh being triangulated
trin.points = t_list[number]
holesR = hs2[['x', 'y', 'ground_height']].values

facesR = np.insert(t_list[number], 0, np.full((1, len(t_list[number])), 3), axis=1)
trinR.faces = facesR
polygonR.lines = linesR
p = pv.Plotter(window_size=[750, 450], notebook=False)#, off_screen=True)
p.add_mesh(trinR, color="blue", show_edges=True, opacity=0.2)
p.add_mesh(polygonR, color="black", opacity=0.3, render_points_as_spheres=False)
p.add_mesh(holesR, color="red")
    
p.set_background('white')
p.show()
```
where number is the position in t_list; I can see the constrained Delaunay within each road segment (I choose to include the aoi to prevent the convex hull, of the road segments, crossing adjacent terrain). 

Not all of t_list is being added as Transport CityObjects. Some are there and some are not. 
What am I missing?

*Let me know if I am asking a bit much and need to give you time. Also if I am over-complicating this and there is an better method; your advice is welcome.**

**Comment by liberostelios** [comment]
Hi @AdrianKriger,

Not sure if this is the best place either, but it's fine at the end of the day.

Have you first checked if the aforementioned objects are in the CityJSON file at all by inspecting its content? It might as well be an issue with ninja not being able to visualise these geometries. Would it be possible to share the specific file shown in the screenshot so that we can check?

**Comment by AdrianKriger** [comment]
SuperSpeedy customer service @liberostelios thank you -> [here](https://github.com/AdrianKriger/osm_LoD1_3DCityModel/blob/main/village_campus/mwe/citjsnClean_cput3d-str.json).

**Comment by liberostelios** [comment, marked as answer]
I don't see the objects in the CityJSON at all. Do you happen to have an easy way of running this script with the given input data? I guess some debugging is needed here.

**Reply by AdrianKriger** [reply]
Thank you @liberostelios; let me have another look at this. 
I was hoping the problem was something basic I overlooked. 

Its strange that some are there and others not. Perhaps plotting all the (triangulated) road segments with the (triangulated) terrain will reveal the challange.

**Reply by AdrianKriger** [reply]
It was my clunky code.

Duplicate id on a road object (after vector manipulation) must have been creating some sought of conflict. Manually increment a new numeral id [here](https://github.com/AdrianKriger/osm_LoD1_3DCityModel/blob/main/village_campus/mwe/osm3DCode2.py#L272) did the business.

![image](https://user-images.githubusercontent.com/59996720/174886388-a04035c5-80f6-4496-858b-6882d0304594.png)

**Reply by liberostelios** [reply]
Awesome! I am glad you tracked down the issue. 🙏
